File: python3/check_if_a_string_contains_all_binary_codes_of_size_k.py
"""
space : O(n)
time  : O(n)

Method:
Sliding window
"""


class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        checklist = set()
        n = len(s)

        for i in range(n-k+1):
            checklist.add(s[i:i+k])

        return len(checklist) == 2 ** k

File: golang/check_if_a_string_contains_all_binary_codes_of_size_k.go
package main

/*
space : O(n)
time  : O(n)

Method:
Sliding window
*/

import (
	"math"
)

func hasAllCodes(s string, k int) bool {
	hashmap := make(map[string]int)
	n := len(s)

	for i := 0; i < (n - k + 1); i++ {
		hashmap[s[i:i+k]] = 1
	}

	return len(hashmap) == int(math.Pow(2, float64(k)))
}
